### ISSUE TITLE
bertQKVToContextPlugin Initialization data

### ISSUE THREAD

**Issue by Qiancj0605** [NVIDIA/TensorRT#929]
![image](https://user-images.githubusercontent.com/34846384/100213404-8a0a9700-2f49-11eb-874f-18a663881a09.png)

How is this data generated？
Can you provide data in "*fp16_312_64*" format? Or can you provide tools to generate data？

**Comment by Qiancj0605** [comment]
@ttyio hi，is there a solution to this problem?

**Comment by ttyio** [comment]
Hello @Qiancj0605 , thank for reporting
these are cubin file encoded in hex buffer, the source code to generate these cubin not open sourced yet. Also the seqlen is selected that align with the tensorcore instructions. 
Could I understand why do you want slen 312? could you padding to 384? Besides that, we also have a variable seqlen implementation, see https://github.com/NVIDIA/TensorRT/tree/release/7.2/demo/BERT#variable-sequence-length

**Comment by Qiancj0605** [comment]
@ttyio I'm sorry, there is a problem with my description. Can you provide data in "fp16_128_52" format?

**Comment by ttyio** [comment]
Hello @Qiancj0605 , currently we only support headsize 64 in demoBERT. It is not customizable.  
Out of curious, which model are you using? Thanks！

**Comment by Qiancj0605** [comment]
@ttyio Thanks。For efficiency reasons, we simplify bert。It’s a pity that the demand cannot be supported。

**Comment by ttyio** [comment]
@Qiancj0605 what do you mean by efficiency?  memory consumption or performance consideration.

**Comment by Qiancj0605** [comment]
@ttyio Mainly for performance considerations。

**Comment by ttyio** [comment]
@Qiancj0605 , have you tried int8 or variable length?

**Comment by Qiancj0605** [comment]
@ttyio thanks，I recently tried int8。

**Comment by Qiancj0605** [comment]
@ttyio When I try int8 by these parameters "--int8 --fp16 --strict -iln",  I encountered this error "[E] SkipLayerNormDQQ - FATAL: unsupported hidden layer size: 384". It's ok If this parameter "-iln" is not used. What could be the reason？

**Comment by ttyio** [comment]
Hello @Qiancj0605 ,
The `SLN` in DemoBert only support bert base (hidden size 768) and bert large (hidden size 1024).

**Comment by Qiancj0605** [comment]
@ttyio Thanks.

**Comment by Qiancj0605** [comment]
@ttyio  hi，In order to support hidden size 384, will I have a problem with writing this way?

![image](https://user-images.githubusercontent.com/34846384/102305453-15040d80-3f9b-11eb-97ff-d95d252e20b5.png)

**Comment by ttyio** [comment]
Hello @Qiancj0605 , we have no test cover custom hidden size. 

You also need train your model in framework, right? My suggestion is that you can try change this, mark skipLayerNorm as output, and compare with the reference value that produced in the framework. If you see mismatch then you have to dig into the CUDA kernel. Thanks!

**Comment by Qiancj0605** [comment]
@ttyio Thanks, I try it.

**Comment by Qiancj0605** [comment]
@ttyio Hi，Through preliminary exploration，the performance of the generated model is greatly improved. 
But I encountered this accidental error(CUDA Error: CUDA_ERROR_INVALID_HANDLE  ../plugin/bertQKVToContextPlugin/fused_multihead_attention.h 269) when using online,What could be the reason？
![image](https://user-images.githubusercontent.com/34846384/109098083-962a0e80-775b-11eb-89f2-19dcb379565e.png)

**Comment by ttyio** [comment]
@Qiancj0605 , are you using multiple gpu in the online service? If so could you try modify the `hashID` function in https://github.com/NVIDIA/TensorRT/blob/master/plugin/bertQKVToContextPlugin/fused_multihead_attention.h#L196

change to:

    inline uint64_t hashID(Data_type type, uint32_t sm) const
    {
        // use deviceID in hasID for multi GPU support before driver support context-less loading of cubin
        int32_t deviceID{0};
        cudaGetDevice(&deviceID);

        ASSERT((deviceID & 0xFFFF) == deviceID);
        ASSERT((type & 0xFFFF) == type);
        ASSERT((sm & 0xFFFFFFFF) == sm);
        return (uint64_t) type << 48 | (uint64_t) deviceID << 32 | sm;
    }

I have commit to internal repo, not integrate to public repo yet.

**Comment by Qiancj0605** [comment]
@ttyio Thanks, i have tried it.But I have to make the following changes，otherwise it will report an error([E] [TRT] safeContext.cpp (184) - Cudnn Error in configure: 7 (CUDNN_STATUS_MAPPING_ERROR)).
Is there any problem with this modification？
![image](https://user-images.githubusercontent.com/34846384/109135406-005da600-7792-11eb-8f29-595c2003554c.png)

**Comment by ttyio** [comment]
Hello @Qiancj0605 , mha plugin would select wrong kernel after comment out this code.  Could you enable verbose log and I want to see where the CUDNN_STATUS_MAPPING_ERROR come from.

BTW, for your deployment environment, Is it possible to share us more information:
- how many gpus in your system?
- how many cuda streams that doing trt inference?
- how is the schedule? like how incoming request mapping to cuda stream?
- how is the async data I/O managed, like when do we sync the cuda stream and copy back data, how do we guarantee there is no race check?

Thanks!

**Comment by Qiancj0605** [comment]
@ttyio I'm so Sorry,there is a problem with my code modification.
There is no problem now, and the result is accurate.I continue to observe it.
Thanks again very much。

**Comment by Qiancj0605** [comment]
@ttyio Hi，After PTQ, the performance is good。But the income has fallen, it should be caused by precision loss(I use one day data to calibrate), is there any way to solve?
The parameters I use include：--int8 --fp16 --strict --calib-num 1000 -iln -imh。

**Comment by ttyio** [comment]
Hello @Qiancj0605 , we also experiment different calibration algorithm for PTQ, and finally find 99.99% percentile max best fit in BERT. We also observe the accuracy loss, that's why we provide a QAT model for bert in

https://github.com/NVIDIA/TensorRT/tree/master/demo/BERT#evaluating-qat-quantization-aware-training-int8-accuracy-using-the-squad-dataset

Currently we have open sourced the QAT toolkit https://github.com/NVIDIA/TensorRT/tree/master/tools/pytorch-quantization, and the opensource for how to generating this BERT QAT model is still in process.

**Comment by Qiancj0605** [comment]
@ttyio Thanks, I  will try it!

**Comment by Qiancj0605** [comment]
@ttyio Recently, I will do QAT quantization by referring to the TF official solution, and more problems in the generation of ONNX.
When will it be released?（the opensource for how to generating this BERT QAT model）Thanks。

**Comment by ttyio** [comment]
Hello @Qiancj0605 , could you take a look into https://github.com/NVIDIA/DeepLearningExamples/tree/master/FasterTransformer/v3.1/bert-quantization/bert-pyt-quantization, we have BERT QAT release by another team in faster-transformer, not sure if this helps you.

Adding @mk-nvidia  @rajeevsrao to comment on the schedule/plan of e2e BERT QAT release in TensorRT. thanks!

**Comment by Qiancj0605** [comment]
Hi @ttyio
A little sorry, it is based on pytorch, which is not effective for us.

**Comment by ttyio** [comment]
Hello @Qiancj0605 , could you check this one? It is TF.
https://github.com/NVIDIA/DeepLearningExamples/tree/master/FasterTransformer/v3.1/bert-quantization/bert-tf-quantization

**Comment by Qiancj0605** [comment]
@ttyio Thanks, I try it.

**Comment by Qiancj0605** [comment]
@ttyio According to this(https://github.com/NVIDIA/DeepLearningExamples/tree/master/FasterTransformer/v3.1/bert-quantization/bert-tf-quantization),get a QAT file in savemodel format.Then did the following changes：
     savemodel→onnx，through this project（https://github.com/onnx/tensorflow-onnx）
     onnx→trt，through this（python3 builder.py -o /workspace/TensorRT/demo/BERT/engines/bert_large_384_int8mix.engine -b 1 -s 384 --int8 --fp16 --strict -c /workspace/TensorRT/demo/BERT/models/fine-tuned/bert_tf_ckpt_large_qa_squad2_amp_384_v19.03.1 -v /workspace/TensorRT/demo/BERT/models/fine-tuned/bert_tf_ckpt_large_qa_squad2_amp_384_v19.03.1/vocab.txt -x /workspace/TensorRT/demo/BERT/models/fine-tuned/bert_pyt_onnx_large_qa_squad11_amp_fake_quant_v1/bert_large_v1_1_fake_quant.onnx -iln -imh.
But I have encountered some problems：“Encountered unknown case”
![image](https://user-images.githubusercontent.com/34846384/113253779-545f2b80-92f8-11eb-8bf7-2a24615dd6f3.png)
What is the problem with this treatment? Thanks

**Comment by ttyio** [comment]
@Qiancj0605 , 
hmmm, the `*_amax` is no longer exported in ONNX file in latest Pytorch, because they are detected as unused variables.
Could you save the model as a pickle file, and modify the code to load the `*_amax` from the pickle file? thanks!

**Comment by Qiancj0605** [comment]
Firstly，I am using TF. Is it the same way with Pytorch?
Secondly，When the model is transformed, how should this ONNX file generate?(python3 builder.py -o ... -b 1 -s 384 --int8 --fp16 --strict -c ...-v ... -x  xx.onnx)。
@ttyio  thanks

**Comment by ttyio** [comment]
@Qiancj0605 , oh sorry. Could you add more debug print in the script, find out which key is missing when the exception is raised? Then try to search from the TF saved model, to see if the key can be found there?
Previous I see some people complain Pytorch has this issue in newer build, not sure TF also has this issue (remove unused variables).

**Comment by ttyio** [comment]
Close since no activation for more than 3 weeks, please reopen if you still have question, thanks!
